### DISCUSSION TITLE
Collision and Weird teleportation problem

### DISCUSSION THREAD

**Discussion by jajaa17** [MonoGame/MonoGame#8403] [General]
Hi everyone im just new to monogame (just a student too making our final project rn, I wanna ask how do we fix this problem thats causing the character to spaz out or twitch or even teleport all of a sudden)

 https://pastebin.com/iCjvCnvZ // player.cs code
 https://pastebin.com/0fxTQmqW // gamestage.cs code

**Comment by trinith** [comment]
Is it the character's position that jumps around? You might need to be more descriptive about the cirumstances that are happening when the issue occurs. Like, are they just standing still or are they on a platform? Is the platform moving? I suspect the issue is somewhere in the code where you handle the collisions.

Also, I don't think this is the problem, but heads up...

```
            // Clamp position to avoid minor variations
            if (currentState == PlayerState.Idle)
            {
                Position = new Vector2((float)Math.Round(Position.X), (float)Math.Round(Position.Y));
            }
```

You shouldn't do this. Keep the precision for your character's position and don't round it. Only round it when you actually draw, which you're already doing here...

```
            Vector2 drawPosition = new Vector2((float)Math.Round(Position.X), (float)Math.Round(Position.Y));
            spriteBatch.Draw(Texture, drawPosition, sourceRectangle, Color.White);
```

You can also simplify that a bit with either of the following options.

```
Vector2 drawPosition = new Vector2((int)Position.X, (int)Position.Y);
Vector2 drawPosition = Position.ToPoint().ToVector2();
```

Anyway, I doubt this is where your problem is. I think you'll have to do some debugging when you get into the bad state. Sorry I couldn't be more helpful!

**Comment by stromkos** [comment]
Most likely the you have the incorrect source/destination offset positions for the sprite atlas you are using.  

Every frame must have a constant reference point that provides continuity among the frames of an animation.

An explosion animation would look odd if it grew from the top left origin of the sprite, only growing down and to the right.
